### PR TITLE
hip: init at 4.3.1

### PR DESCRIPTION
**Pull request by lovesegfault**
###### Motivation for this change
Follow-up from https://github.com/nixos-rocm/nixos-rocm/issues/65

cc. @acowley

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by acowley** [comment]
Great to see this! Why version 4.0.0? It looks like the ROCm llvm here is at 4.1, and the nixos-rocm overlay is at 4.3.

**Comment by lovesegfault** [comment]
> Great to see this! Why version 4.0.0? It looks like the ROCm llvm here is at 4.1, and the nixos-rocm overlay is at 4.3.

Just because it's a 1-1 transplant from the code in `nixos-rocm`. I'm going to bump the versions now that I got the baseline working :)

**Comment by acowley** [comment]
Okay, let me know if you have any trouble. The overlay is working fine at 4.3 afaik.

**Comment by lovesegfault** [comment]
Alright, @acowley, I bumped our entire ROCm ecosystem to 4.3.1 and the new packages as well :)

**Comment by lovesegfault** [comment]
Result of `nixpkgs-review pr 136445` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>17 packages built:</summary>
  <ul>
    <li>hip</li>
    <li>llvmPackages_rocm.clang</li>
    <li>llvmPackages_rocm.clang-unwrapped</li>
    <li>llvmPackages_rocm.clangNoCompilerRt</li>
    <li>llvmPackages_rocm.compiler-rt</li>
    <li>llvmPackages_rocm.lld</li>
    <li>llvmPackages_rocm.llvm</li>
    <li>rocclr</li>
    <li>rocm-cmake</li>
    <li>rocm-comgr</li>
    <li>rocm-device-libs</li>
    <li>rocm-opencl-icd</li>
    <li>rocm-opencl-runtime</li>
    <li>rocm-runtime</li>
    <li>rocm-smi</li>
    <li>rocm-thunk</li>
    <li>rocminfo</li>
  </ul>
</details>

**Comment by lovesegfault** [comment]
@acowley If you could take a look at this I think it's ready to merge, I'd like a second opinion before hitting the button though :)

**Comment by risicle** [comment]
Is there any way to test/review this without hardware?

**Comment by lovesegfault** [comment]
> Is there any way to test/review this without hardware?

Not as far as I can tell, I have an AMD GPU and did some basic testing locally, where things looked fine.

File: pkgs/development/compilers/hip/default.nix [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+  src = fetchFromGitHub {
+    owner = "ROCm-Developer-Tools";
+    repo = "HIP";
+    rev = "rocm-${version}";
+    sha256 = "sha256-dUdP32H0u6kVItS+VUE549vvxkV1mSN84HvyfeK2hEE=";
+  };
+
+  # FIXME: https://github.com/ROCm-Developer-Tools/HIP/issues/2317
+  postPatch = ''
+    cp ${rocm-opencl-runtime.src}/amdocl/cl_vk_amd.hpp amdocl/
```
Comment (acowley) [inline pkgs/development/compilers/hip/default.nix:38]: I don't have this fix in the nixos-rocm overlay, but 4.3.0 is building fine. Any idea?

File: pkgs/development/compilers/hip/default.nix [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+    "-DHIP_CLANG_ROOT=${clang-unwrapped}"
+  ];
+
+  # patches = [(fetchpatch {
+  #   # See https://github.com/ROCm-Developer-Tools/HIP/pull/2005
+  #   name = "hiprtc-fix-PR2005";
+  #   url = "https://patch-diff.githubusercontent.com/raw/ROCm-Developer-Tools/HIP/pull/2005.patch";
+  #   sha256 = "1w35s2xpxny4j5llpaz912g1br9735vdfdld1nhqdvrdax2vxlc7";
+  # })];
+
```
Comment (acowley) [inline pkgs/development/compilers/hip/default.nix:83]: Let's delete that commented out `patches` definition.

File: pkgs/development/compilers/hip/default.nix (hunk 1 above)
Comment (lovesegfault) [reply pkgs/development/compilers/hip/default.nix:38]: Well, this is 4.3.1 so perhaps the issue happened on the 0 -> 1 patch changes?

File: pkgs/development/compilers/hip/default.nix (hunk 2 above)
Comment (lovesegfault) [reply pkgs/development/compilers/hip/default.nix]: Done, nice catch!

File: pkgs/development/compilers/hip/default.nix (hunk 1 above)
Comment (lovesegfault) [reply pkgs/development/compilers/hip/default.nix:38]: Interesting, even for 4.3.0 the Gentoo package needs this: https://github.com/gentoo/gentoo/blob/ab005b2b0219f266a90891fa3cce8253014927db/dev-util/hip/hip-4.3.0.ebuild#L69

Comment (lovesegfault) [reply pkgs/development/compilers/hip/default.nix:38]: @acowley Am I cleared to merge this even with this fix?

Comment (acowley) [reply pkgs/development/compilers/hip/default.nix:38]: Sure!

**Comment by lovesegfault** [comment]
:partying_face:

**Pull request merged**